Create yaml config design guidance (#2159)
apollographql/router: dev-docs/yaml-design-guidance.md

5. [Plan for the future](#plan-for-the-future).

### Avoid empty config

In Rust you can use `Option` to say that config is optional, however this can give a bad experience if the type is complex and all fields are optional.

#### GOOD
```rust
#[serde(deny_unknown_fields)]
struct Export {
    url: Url // url is required
}
```
```yaml
export:
    url: http://example.com
```

#### GOOD
```rust
enum ExportUrl {
    Default,
    Url(Url)
}

#[serde(deny_unknown_fields)]
struct Export {
    url: ExportUrl // Url is required but user may specify `default`
}
```
```yaml
export:
    url: default
```

#### BAD
```rust
#[serde(deny_unknown_fields)]
struct Export {
    url: Optional<Url> // url is optional
}
```
```yaml
export: # The user is not aware that url was defaulted.
```

#### UGLY
In the case where you genuinely have no config then it may be acceptable to have an `enabled: bool` flag.
The reason to avoid this is that it creates a disconnect between the user ant the thing that they are trying to do
```rust
#[serde(deny_unknown_fields)]
struct Export {
    enabled: bool,
    url: Optional<Url> // url is optional

> }

Reviewer: I think sometimes it's not so ugly in case you have a default value for the optional value in yaml. You should mention that.
Let me explain why it's not ugly/bad when you have a default value.
+ Having to write always the default value when you just want to enable something could be boring.
+ Enforcing this default value could also be very useful in specific cases instead of asking to the users to write it by themselves just to enable something. The main reason is to have a kind of convention. Like if we know there is a default value, let's say a default header name. We can do some assumptions on that, for example if you're developing a graphql client which wants to be by default compliant with the router, it knows the default value and are able to be compliant for free, it's like a convention because it's the default value.

Author: The problem is the lack of consistency. There are lots of places in the config we could have an `enabled` field. Pretty much at every level

For instance telemetry export:
```yaml
telemetry:
  enabled: true # Enable or disable telemetry plugin completely
  tracing:
    enabled: true # Enable or disable tracing
    datadog: 
      enabled: true # Defaults to datadog agent at localhost. Endpoint can become optional now
    otlp: 
      enabled: true # Defaults to GRPC to localhost, Endpoint can become optional now
```

If we decide to go down the enabled route we should look at the rest of the config and put `enabled` in so that the user gets a consistent experience.

There is an advantage to using enabled across the board. If users want to enable or disable stuff via env variable they can, but I'm not completely convinced.

Author: This is the pattern we have used elsewhere:
https://github.com/apollographql/router/pull/2159/files/9093b125f43d3fe19ca3d5b28c055705b6f89a94#diff-2897bc49cf0aab22634663d6ae15fd678e61323c9a5a63e5a90003a9d0973bbdR60-R75

Author: Changed in
[2f3193a](https://github.com/apollographql/router/pull/2159/commits/2f3193a84cda2159342a1b4eceea68e0f6644f5e)

```
```yaml
export: 
  enabled: true # The user is not aware that url was defaulted.
```

### Do use `#[serde(deny_unknown_fields)]`.
Every container that takes part in config should be annotated with `#[serde(deny_unknown_fields)]`. If not the user can make mistakes on their config and they they won't get errors.

#### GOOD
```rust
#[serde(deny_unknown_fields)]
struct Export {
    url: Url
}
```
```yaml
export: 
  url: http://example.com
  backup: http://example2.com # The user will receive an error for this
```

#### BAD
```rust
struct Export {
    url: Url
}
```
```yaml
export: 
  url: http://example.com
  backup: http://example2.com # The user will NOT receive an error for this
```

### Don't use `#[serde(flatten)]`
Serde flatten is tempting to use where you have identified common functionality, but creates a bad user experience as it is incompatible with `#[serde(deny_unknown_fields)]`. There isn't a great solution to this, but you can use a macro to make things dry.

See [serde documentation](https://serde.rs/field-attrs.html#flatten) for more details.

#### UGLY
```rust
#[serde(deny_unknown_fields)]
struct Export {

>     common_export_fields!()

Reviewer: Could you provide an example of what look like that kind of macro ?

Author: Hah, it's impossible. I'll change the advice to just expand all the fields.

Author: [02478bb](https://github.com/apollographql/router/pull/2159/commits/02478bbb4edff308e384e17e8927be2a436f8d6e)

# Yaml config design

The router uses yaml configuration, and when creating new features or extending existing features you'll likely need to think about how configuration is exposed.

In general users should have a pretty good idea of what a configuration option does without referring to the documentation.

## Migrations

We won't always get things right, and sometimes we'll need to provide [migrations](apollo-router/src/configuration/migrations/README.md) from old config to new config.

Make sure you:
1. Mention the change in the changelog
2. Update docs
3. Update any test configuration
4. Create a migration test as detailed in [migrations](apollo-router/src/configuration/migrations/README.md)
5. In your migration description tell the users what they have to update.

## Process
It should be obvious to the user what they are configuring and how it will affect Router behaviour. It's tricky for us as developers to know when something isn't obvious to users as often we are too close to the domain.

> Complex config should be run by the rest of the team. Ideally before writing any code, as coming to the team late can cause code churn and frustration. The process is as follows:

Reviewer 2: Maybe let's have a less negative outlook here

```suggestion
Complex configuration changes should be discussed with the team before starting the implementation, since they will drive the code's design. The process is as follows:
```

1. In the github issue put the proposed config in.
2. List any concerns.
3. Notify the team that you are looking for request for comment.

> 4. Schedule a meeting to discuss. (This is important, often design considerations will fall out of conversation)

Reviewer 2: is this doc only for the router team, or would that be a reference for future contributors too? (that would have a harder time scheduling meetings with us)

Author: [14404db](https://github.com/apollographql/router/pull/2159/commits/14404db4744533ef697c7d4de62677ffb14c68d9)

5. If it is not completely clear what the direction should be:

> 1. Wait a few days, often people will have ideas later even if they didn't in the meeting.
> 2. Ask users what they think.

Reviewer 2: ```suggestion
  1. Wait a few days, often people will have ideas later even if they didn't in the meeting.
  2. Ask users what they think.
```

Reviewer 2: maybe we should ask users earlier in the process

Author: [4c06e20](https://github.com/apollographql/router/pull/2159/commits/4c06e20db6bbd63982f8364e9b17549ee36bb889)

    url: Optional<Url> // url is optional
}
```
```yaml
export: # The user is not aware that url was defaulted.
```

#### UGLY
In the case where you genuinely have no config then it may be acceptable to have an `enabled: bool` flag.
The reason to avoid this is that it creates a disconnect between the user and the thing that they are trying to do
```rust
#[serde(deny_unknown_fields)]
struct Export {
    enabled: bool,
    url: Optional<Url> // url is optional
}
```
```yaml
export: 
  enabled: true # The user is not aware that url was defaulted.
```

### Do use `#[serde(deny_unknown_fields)]`.
Every container that takes part in config should be annotated with `#[serde(deny_unknown_fields)]`. If not the user can make mistakes on their config and they they won't get errors.

#### GOOD
```rust
#[serde(deny_unknown_fields)]
struct Export {
    url: Url
}
```
```yaml
export: 
  url: http://example.com
  backup: http://example2.com # The user will receive an error for this
```

#### BAD
```rust
struct Export {
    url: Url
}
```
```yaml
export: 
  url: http://example.com
  backup: http://example2.com # The user will NOT receive an error for this
```

### Don't use `#[serde(flatten)]`

> Serde flatten is tempting to use where you have identified common functionality, but creates a bad user experience as it is incompatible with `#[serde(deny_unknown_fields)]`. There isn't a great solution to this, but nesting config can sometimes help.

Reviewer 2: that's an interesting point, I did not know about this. Should we remove the current uses of `serde(flatten)`?

See [serde documentation](https://serde.rs/field-attrs.html#flatten) for more details.

#### MAYBE
```rust
#[serde(deny_unknown_fields)]
struct Export {
    url: Url,
    backup: Url
}
#[serde(deny_unknown_fields)]
struct Telemetry {
    export: Export
}
#[serde(deny_unknown_fields)]
struct Metrics {
    export: Export
}
```
```yaml
telemetry:
  export: 
    url: http://example.com
    backup: http://example2.com
metrics:
  export:
    url: http://example.com
    backup: http://example2.com
```

#### BAD
```rust
#[serde(deny_unknown_fields)]
struct Export {
    url: Url,
    backup: Url
}
struct Telemetry {
    export: Export
}
```
```yaml
telemetry: 
  url: http://example.com
  backup: http://example2.com
  unknown: sadness # The user will NOT receive an error for this
```

### Use consistent terminology
Be consistent with the rust API terminology.
* request - functionality that modifies the request or retrieves data from the request of a service.
* response - functionality that modifies the response or retrieves data from the response of a service.

> * supergraph - functionality within Plugin::subgraph_service

Reviewer 2: ```suggestion
* supergraph - functionality within Plugin::supergraph_service
```

* execution - functionality within Plugin::execution_service
* subgraph(s) - functionality within Plugin::subgraph_service

> If you use the above terminology then changes are you are doing something that will take place on every request. In this case make sure to include an `action` verb so the user know what the config is doing.

Reviewer 2: ```suggestion
If you use the above terminology then chances are you are doing something that will take place on every request. In this case make sure to include an `action` verb so the user know what the config is doing.
```

#### GOOD
```yaml
headers:
  subgraphs: # Modifies the subgraph service  
    products: 
      request: # Retrieves data from the request
        - propagate: # The action.
            named: foo
```

#### BAD
```yaml
headers:
  named: foo # From where, what are we doing, when is it happening?
```

### Document your configuration options

> If your config is well documented in Rust then it will be well documented in the generated Json Schema. This means that when users are modifying their config either in their IDE or in Apollo GraphOS documentation is available.

Reviewer 2: ```suggestion
If your config is well documented in Rust then it will be well documented in the generated JSON Schema. This means that when users are modifying their config either in their IDE or in Apollo GraphOS, documentation is available.
```

Example configuration should be included on all containers.

#### GOOD
```rust
/// Export the data to the metrics endpoint
/// Example configuration:
/// ```yaml
/// export:
///   url: http://example.com
/// ```
#[serde(deny_unknown_fields)]
struct Export {
    /// The url to export metrics to.
    url: Url
}
```

#### BAD
```rust
#[serde(deny_unknown_fields)]
struct Export {
    url: Url
}
```

> In addition, make sure to update the html documentation. 

Reviewer 2: ```suggestion
In addition, make sure to update the published documentation in the `docs/` folder.
```

1. [Avoid empty config](#avoid-empty-config).
2. [Do use `#[serde(deny_unknown_fields)]`](#do-use-serdedeny_unknown_fields).
3. [Don't use `#[serde(flatten)]`](#dont-use-serdeflatten).
4. [Use consistent terminology](#use-consistent-terminology).
5. [Don't use negative options](#dont-use-negative-options).
6. [Document your configuration options](#document-your-configuration-options).
7. [Plan for the future](#plan-for-the-future).

### Avoid empty config

In Rust you can use `Option` to say that config is optional, however this can give a bad experience if the type is complex and all fields are optional.

#### GOOD
```rust
#[serde(deny_unknown_fields)]
struct Export {
    url: Url // url is required
}
```
```yaml
export:
    url: http://example.com
```

#### GOOD
```rust
enum ExportUrl {
    Default,
    Url(Url)
}

#[serde(deny_unknown_fields)]
struct Export {
    url: ExportUrl // Url is required but user may specify `default`
}
```
```yaml
export:
    url: default
```

#### GOOD
In the case where you genuinely have no config or all sub-options have obvious defaults then use an `enabled: bool` flag.
```rust
#[serde(deny_unknown_fields)]
struct Export {
    enabled: bool,

>     url: Optional<Url> // url is optional

Reviewer: Is it intended that you use `Optional<>` instead of `Option<>` ?
Also I'm wondering if in that case if it has a default value we should force people to not use optional field but use the field with a serde attribute `#[serde(default = "default_resource")]` What do you think ?

Author: It depends, `default_resource` is OK if the answer is static, not it if is something like an env variable as generating the schema will leak the value.
In this case it's probably OK. I'll update and then merge.
